### DISCUSSION TITLE
Support updating version-specific info inside compatibility suffixes

### DISCUSSION THREAD

**Discussion by wrslatz** [renovatebot/renovate#24849] [Suggest an Idea]
### What would you like Renovate to be able to do?

Renovate should be able to analyze version-specific info (major, minor, patch) within the `compatibility` section of the version of a dependency.

In Docker tags, it is common for the `compatibility` part of the tag to include version info (i.e. `-alpine3.14`). Without this capability, `Dockerfile` tags that include version info in the `compatibility` section, where that same compatibility tag no longer exists in newer versions (e.x. `img1.2-alpine3.14` to `img1.3-alpine3.15`, where `img1.3-alpine3.14` never exists), will never see version updates proposed by Renovate.

Practical example: Node.js images (see current tags [here](https://hub.docker.com/_/node)). It would be great to get proposed updates from any of `17-alpine3.14`, `17.3-alpine3.14`, `17.3.1-alpine3.14` to `17-alpine3.15`, `17.3-alpine3.15`, `17.3.1-alpine3.15` respectively. You could use `-alpine` tags alone, but you lose precision into the underlying Alpine version and increase the risk of consuming breaking changes without knowing.

See the [Docker Version compatibility docs](https://docs.renovatebot.com/docker/#version-compatibility) for more info on how this works today. Regex versioning is also related.

I split this out as a separate request from https://github.com/renovatebot/renovate/issues/5990#issuecomment-1015453454 to be discussed on its own. I'm not sure if this is more relevant to managers or to versioning support (probably the latter?).

### If you have any ideas on how this should be implemented, please tell us here.

Introduce a concept of `compatibilityVersioning`, where the `compatibility` section of the version is parsed for version info. The matching logic could be similar to the general regex matching logic but applied within the `compatibility` section, so something like `^(?<compatibility>.[A-Za-z] )(?<major>\\d+)\\.(?<minor>\\d+)\\.(?<patch>\\d+)?$`.

This would complicate how Renovate looks for updated dependency versions, increasing the number of tags it would need to check. Example: Renovate would need to check combinations of 1) same version info except `compatibility` version update 2) update to version info except same `compatibility` version, and 3) both update to version info and `compatibility` version.

It may also be difficult to parse version info out of `compatibility` (there may be examples where numeric characters are used as part of the `compatibility` tag and do not represent version info). Consumers may also want to separate regular version updates from `compatibility` version updates when applicable, adding additional config and support requirements (similar to `separateMinorPatch`).

Perhaps a preferred order could be considered here to make the proposed updates more deterministic. Ex: Renovate could look for updates to both version and compatibility, version alone, and then compatibility alone.

Some alternatives or workarounds that may be suitable
1) Use a tag with less-specific `compatibility` (i.e. `-alpine` over `-alpine3.14`) in combination with digest pinning. This would help prevent consuming breaking changes from compatibility updates in the less specific tag. It still reduces visibility into the underlying Alpine version being used.
2) Use replacements to manually replace specific instances where the specific `compatibility` version info will no longer receive updates. This would depend on Docker support in replacements being completed in https://github.com/renovatebot/renovate/issues/12623 first. It's also a manual process for each image, making it less than ideal and a large burden to maintain for many images.

### Is this a feature you are interested in implementing yourself?

Maybe

**Comment by tchupp** [comment]
I haven't seen any conversation on this suggestion, but I wanted to give it a +1 👍🏽

The NodeJS example is very relevant for my company.

Lemme talk out an example;
If a Dockerfile has `FROM node:18-alpine3.17`, Renovate currently extracts:
```
{
          autoReplaceStringTemplate:
            '{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}',
          currentDigest: undefined,
          currentValue: '18-alpine3.17',
          datasource: 'docker',
          depName: 'node',
          depType: 'final',
          replaceString: 'node:18-alpine3.17',
        }
```
Renovate might find three potential upgrades:
- `18-alpine3.18`
- `20-alpine3.17`
- `20-alpine3.18`

I suppose, based on settings, we would expect to see either `18-alpine3.18` or `20-alpine3.17` as the next upgrade that Renovate proposed.

**Comment by der-ali** [comment]
+1

**Comment by ghost** [comment]
Hello,
Some news about this feature ? 
Is there a bypass to implement the upgrade version on compatibility section ?
Thanks

**Comment by elchenberg** [comment]
I would like to see this feature for Debian release names, too, for example in container image versions `node:23.3.0-bullseye-slim` could be updated to `node:23.3.0-bookworm-slim`.

I tried to use a packagesRule with replacements and the `replace` template function to replace the Debian version names in any image tag but this does not work because the `replacementVersion` is not one of the fields that is passed through for templating (see <https://docs.renovatebot.com/templates/#exposed-config-options>):

```json
    "packageRules": [
        {
            "matchDatasources": [
                "docker"
            ],
            "matchCurrentValue": "/-?bullseye-?/",
            "replacementVersion": "{{{replace 'bullseye' 'bookworm' currentValue}}}"
        }
    ]
```

But if you have a specific image version it does work for, for example this is a configuration is used for a Golang image:

```js
    {
      matchDatasources: [
        "docker"
      ],
      matchDepNames: [
        "golang",
        "library/golang",
        "docker.io/library/golang",
        "public.ecr.aws/docker/library/golang"
      ],
      matchCurrentValue: "1.22.4-bullseye",
      replacementVersion: "1.22.4-bookworm"
    }
```

In some places I used custom.regex managers to update Debian release names and Alpine versions in container tags but it would be much nicer if I could use packageRules instead to avoid duplicating and then maintaining the `fileMatch` and `matchStrings` in multiple places:

```json
    "customManagers": [
        {
            "description": "Update Debian version codenames in container image tags, for example 'buster' in 'node:20.10.0-buster'.",
            "customType": "regex",
            "fileMatch": [
                "(^|/|\\.)([Dd]ocker|[Cc]ontainer)file$",
                "(^|/)([Dd]ocker|[Cc]ontainer)file[^/]*$",
                "\\.ya?ml$"
            ],
            "matchStrings": [
                "(?:^|[ \"'])(?<depName>[\\w/.:-]+?):[\\w.-]+?-(?<currentValue>[a-z]+)(?:-slim)?(?:@(?<currentDigest>sha256:[0-9a-f]{64}))?(?:$|[ \"'])"
            ],
            "datasourceTemplate": "docker",
            "versioningTemplate": "debian"
        },
        {
            "description": "Update the Alpine version in container image tags. For example, update php:8.2.24-fpm-alpine3.19 to php:8.2.24-fpm-alpine3.20.",
            "customType": "regex",
            "fileMatch": [
              "(^|/|\\.)([Dd]ocker|[Cc]ontainer)file$",
              "(^|/)([Dd]ocker|[Cc]ontainer)file[^/]*$",
              ".*\\.ya?ml$"
            ],
            "matchStrings": [
              "(?:FROM|image:)[ ]+[\"']?(?<depName>[^@\"'\\s]+):(?<currentValue>[^:@\"'\\s]*alpine\\d+\\.\\d+[^:@\"'\\s]*)(?:@(?<currentDigest>sha256:[0-9a-f]+))?[\"']?\\s"
            ],
            "versioningTemplate": "regex:^(?<compatibility>.*alpine)(?<major>\\d+)\\.(?<minor>\\d+).*$",
            "datasourceTemplate": "docker"
        }
    ]
```

For Alpine I also used this package rule as a workaround, where the Alpine major and minor versions are treated as build and revision version:

```json
    {
      "description": "Update Alpine version in container image tags like 'docker.io/library/node:21.6.1-alpine3.18'.",
      "matchDatasources": [
        "docker"
      ],
      "matchCurrentValue": "/^(?:(?<major>\\d+)(?:\\.(?<minor>\\d+)(?:\\.(?<patch>\\d+))?)?-)?alpine(?<build>\\d+)(?:\\.(?<revision>\\d+))?$/",
      "versioning": "regex:^(?:(?<major>\\d+)(?:\\.(?<minor>\\d+)(?:\\.(?<patch>\\d+))?)?-)?alpine(?<build>\\d+)(?:\\.(?<revision>\\d+))?$"
    },
```

**Reply by gaeljw** [reply]
For the record, I've also built a custom manager for Ubuntu-based images like `eclipse-temurin`: https://gaeljw.github.io/renovate_preset_docker_eclipse_temurin_os_variant.html (the `packageRule` piece is specific to the fact it's `eclipse-temurin`, it's not useful for other images).

More specific than yours but sharing it in case it helps someone:
```json5
{
      "customType": "regex",
      // In which file to look for something to do
      // Same pattern as standard Docker manager (https://docs.renovatebot.com/modules/manager/dockerfile/)
      "managerFilePatterns": [
        "/(^|/|\\.)([Dd]ocker|[Cc]ontainer)file$/",
        "/(^|/)([Dd]ocker|[Cc]ontainer)file[^/]*$/"
      ],
      // Regex to extract some attributes from the Dockerfile
      // I.e. from "FROM eclipse-temurin:21.0.5_11-jre-jammy", how to extract "21.0.5_11" and "jammy"
      "matchStrings": [
        "FROM eclipse-temurin:(?<regularVersion>.*)-jre-(?<currentValue>.*)"
      ],
      // From the list of available versions (image tags on container registry), how to extract versions
      // i.e. from "21.0.5_11-jre-noble" or "21.0.6_7-jre-noble", how to filter only the same Java version and identify
      // "noble" as the version we care about here
      "extractVersionTemplate": "^{{{regularVersion}}}-jre-(?<version>.*)$",
      // Defining a custom depName so that it doesn't clash with regular updates of Java version
      "depNameTemplate": "eclipse-temurin-os-variant",
      // Name of the package to look for in the datasource
      // Using packageName because depName is overriden
      "packageNameTemplate": "eclipse-temurin",
      // Versioning to use
      // This is where the magic appear: Renovate knows that noble > jammy for instance
      "versioningTemplate": "ubuntu",
      // Where to look for available versions
      "datasourceTemplate": "docker"
    }
```
